diff --git a/.gitignore b/.gitignore
index 518c37d..4279eae 100644
--- a/.gitignore
+++ b/.gitignore
@@ -10,3 +10,7 @@ src/build.log
 # Ignore vscode settings
 .vscode
 .vscode/**
+.vscode-ctags
+
+build_*/**
+*.img.*
\ No newline at end of file
diff --git a/build.sh b/build.sh
new file mode 100644
index 0000000..0d24684
--- /dev/null
+++ b/build.sh
@@ -0,0 +1,223 @@
+#!/bin/bash
+# Generations of developers used file hierarchies instead of "yaml archives"
+set -e
+
+TARGET_NAME=${1:-"snapmakerj1"}
+IMAGE_TYPE=${2:-armbian}
+SPECIAL_MODULES=${3:-"50-kiauh"}
+
+START_TIME=$(date +'%Y%m%d_%H%M%S')
+WORKDIR=build_${TARGET_NAME}_${START_TIME}
+
+if [ -e "${WORKDIR}" ]; then 
+    echo Workdir "'${WORKDIR}'" already exists 1>&2
+    exit 1
+fi 
+
+mkdir -p "${WORKDIR}"
+
+#Unpack image 
+echo "🛠 Unpack and rename image"
+
+# search for the image file in the build directory
+IMG_FILE=$(ls -t *.img *.img.xz 2>/dev/null | head -n 1)
+
+if [[ "$IMG_FILE" == *.img.xz ]]; then
+    echo "Found compressed Image file: ${IMG_FILE}"
+
+    # extract the image file
+    xzcat -T0 "${IMG_FILE}" > "${WORKDIR}/input.img"
+elif [[ "$IMG_FILE" == *.img ]]; then
+    echo "Found uncompressed image file: ${IMG_FILE}"
+    ln "${IMG_FILE}" "${WORKDIR}/input.img"
+else
+    echo "No images found... " 1>&2
+    exit 1
+fi
+
+# remove disabled
+function remove_disabled
+{
+    local path=${1}
+
+    if [ ! -d "${path}" ]
+    then 
+        echo Path "'${path}'" is not a directory 1>&2
+        exit 1
+    fi
+
+    for file in "${path}"/*.disabled
+    do     
+        if [ -e "${file}" ]
+        then 
+            module=${path}/$(basename "${file}" .disabled)
+            rm -f "${module}"
+            rm -f "${file}"
+        fi
+    done 
+}
+
+function patch_append
+{
+    local targetpath=${1}
+    
+    for appendfile in "${targetpath}"/*.append
+    do 
+        if [ -f "${appendfile}" ]
+        then 
+            module=${targetpath}/$(basename "${appendfile}" .append)
+            if [ -f "${module}" ]
+            then
+                cat "${appendfile}" >> "${module}"
+            fi
+            rm "${appendfile}"
+        fi
+    done 
+}
+
+function copy_files
+{
+    local source_path=${1}
+    local target_path=${2}
+
+    if [ ! -d "${source_path}" ]
+    then 
+        return 0
+    fi
+
+    if [ ! -e "$target_path" ]
+    then 
+        mkdir -p "${target_path}"
+    fi
+
+    for file in "${source_path}"/*
+    do
+        if [ -f "${file}" ]
+        then 
+            target_file=${target_path}/$(basename "${file}")
+            cp "${file}" "${target_file}"
+        fi
+    done
+}
+
+function copy_files_recursive
+{
+    local source_path=${1}
+    local target_path=${2}
+
+    if [ ! -d "${source_path}" ]
+    then 
+        return 0
+    fi
+
+    if [ ! -e "$target_path" ]
+    then 
+        mkdir -p "${target_path}"
+    fi
+
+    for file in "${source_path}"/*
+    do
+        if [ -e "${file}" ]
+        then 
+            target_file=${target_path}/$(basename "${file}")
+            cp -r "${file}" "${target_file}"
+        fi
+    done
+}
+
+# Copy generic scripts 
+echo "📋 Copy generic and type scripts"
+
+copy_files "modules/generic" "${WORKDIR}/scripts"
+copy_files_recursive "modules/generic/files" "${WORKDIR}/scripts/files"
+#TODO: this recursive
+remove_disabled "${WORKDIR}/scripts"
+remove_disabled "${WORKDIR}/scripts/files"
+patch_append "${WORKDIR}/scripts"
+patch_append "${WORKDIR}/scripts/files"
+
+# copy distro files 
+copy_files "modules/${IMAGE_TYPE}" "${WORKDIR}/scripts/"
+copy_files_recursive "modules/${IMAGE_TYPE}/files" "${WORKDIR}/scripts/files"
+remove_disabled "${WORKDIR}/scripts/"
+remove_disabled "${WORKDIR}/scripts/files"
+patch_append "${WORKDIR}/scripts"
+patch_append "${WORKDIR}/scripts/files"
+
+echo "📋 Copy special modules"
+# YAML.. JSON 
+# why via temp dir??
+echo "Special modules found for ${TARGET_NAME}: ${SPECIAL_MODULES}"
+
+# copy additional files for special modules
+if [ -d "modules/special/files" ]; then
+    echo "Copying common special files..."
+    cp -r modules/special/files/* "${WORKDIR}/scripts/files/"
+fi
+
+for module in "${SPECIAL_MODULES[@]}"
+do 
+    echo "Copying module: ${module}"
+    cp "modules/special/${module}" "${WORKDIR}/scripts/"
+done 
+
+remove_disabled "${WORKDIR}"/scripts
+remove_disabled "${WORKDIR}"/scripts/files
+patch_append "${WORKDIR}/scripts"
+patch_append "${WORKDIR}/scripts/files"
+
+
+copy_files "configs/${TARGET_NAME}/scripts" "${WORKDIR}/scripts/"
+copy_files_recursive "configs/${TARGET_NAME}/scripts/files" "${WORKDIR}/scripts/files"
+remove_disabled "${WORKDIR}/scripts/"
+remove_disabled "${WORKDIR}/scripts/files"
+patch_append "${WORKDIR}/scripts"
+patch_append "${WORKDIR}/scripts/files"
+
+
+echo "📇 Generate config.local"
+if [ -f "configs/${TARGET_NAME}/config.local" ]; then 
+    cp "configs/${TARGET_NAME}/config.local" "${WORKDIR}/config.local"
+else 
+    touch "${WORKDIR}/config.local"
+fi
+
+#EDITBASE_ seems not to be used
+# print("Generating config.local file...")
+# import os, json
+# env_vars = json.loads(os.environ.get("MATRIX_ENV", "{}"))
+# with open("config.local", "w") as f:
+#     for key, value in env_vars.items():
+#         if not key.startswith("EDITBASE_"):
+#             key = "EDITBASE_" + key
+#         f.write(f"{key}={value}\n")
+
+echo "🔎 Show scripts"
+ls -al "${WORKDIR}/scripts"
+
+echo "🏗 Run CustoPiZer"
+
+# if [ -d "./build" ]; then 
+#     rm -rf build
+# fi 
+# mkdir build
+cd "${WORKDIR}"
+docker run --rm \
+    --user 0:0 \
+    --cap-add setuid --cap-add=SYS_PTRACE --privileged  --security-opt seccomp=unconfined --security-opt apparmor=unconfined \
+    -v "${PWD}":/CustoPiZer/workspace \
+    -v "${PWD}/config.local":/CustoPiZer/config.local:ro \
+    custopizer:evil
+#     ghcr.io/octoprint/custopizer:main
+
+# TODO
+exit 0
+
+echo "✏ Rename image"
+IMAGE_NAME="MainsailOS-${IMAGE_TYPE}_${TARGET_NAME}_${START_DATE}.img
+mv output.img "${IMAGE_NAME}"
+
+echo "🛠 Compressing Image"
+echo "🔐 Create hash"
+
+
diff --git a/configs/snapmakerj1/config.local b/configs/snapmakerj1/config.local
new file mode 100644
index 0000000..fb5c964
--- /dev/null
+++ b/configs/snapmakerj1/config.local
@@ -0,0 +1,7 @@
+# source this
+EDITBASE_ARCH=armhf
+EDITBASE_DISTRO=armbian
+EDITBASE_IMAGE_ENLARGEROOT=3000
+EDITBASE_MOUNT_PROC=1
+EDITBASE_MOUNT_SYS=1
+
diff --git a/configs/snapmakerj1/scripts/01-remove-packages b/configs/snapmakerj1/scripts/01-remove-packages
new file mode 100644
index 0000000..9506492
--- /dev/null
+++ b/configs/snapmakerj1/scripts/01-remove-packages
@@ -0,0 +1,19 @@
+#!/bin/bash
+set -xe
+export LC_ALL=C
+
+########################################
+# Functions and Basic Configuration    #
+########################################
+
+# Load common functions and set up cleanup trap
+source /common.sh
+install_cleanup_trap
+
+# Load additional configuration (e.g., BASE_USER)
+source /files/00-config
+
+# save some space
+apt autoremove -y armbian-firmware pciutils
+
+# EOF
\ No newline at end of file
diff --git a/configs/snapmakerj1/scripts/32-canbus.disabled b/configs/snapmakerj1/scripts/32-canbus.disabled
new file mode 100644
index 0000000..e69de29
diff --git a/configs/snapmakerj1/scripts/50-klipper.append b/configs/snapmakerj1/scripts/50-klipper.append
new file mode 100644
index 0000000..3e99ae8
--- /dev/null
+++ b/configs/snapmakerj1/scripts/50-klipper.append
@@ -0,0 +1,5 @@
+# dirty cleanup
+#Not needed on snapmakerj1
+apt autoremove -y avrdude gcc-avr binutils-avr avr-libc
+
+#EOF
\ No newline at end of file
diff --git a/configs/snapmakerj1/scripts/53-crowsnest.disabled b/configs/snapmakerj1/scripts/53-crowsnest.disabled
new file mode 100644
index 0000000..e69de29
diff --git a/configs/snapmakerj1/scripts/54-timelapse.disabled b/configs/snapmakerj1/scripts/54-timelapse.disabled
new file mode 100644
index 0000000..e69de29
diff --git a/configs/snapmakerj1/scripts/56-klipperscreen b/configs/snapmakerj1/scripts/56-klipperscreen
new file mode 100644
index 0000000..c4fedd7
--- /dev/null
+++ b/configs/snapmakerj1/scripts/56-klipperscreen
@@ -0,0 +1,38 @@
+#!/bin/bash
+set -xe
+export LC_ALL=C
+
+########################################
+# Functions and Basic Configuration    #
+########################################
+
+# Load common functions and set up cleanup trap
+source /common.sh
+install_cleanup_trap
+
+# Load additional configuration (e.g., BASE_USER)
+source /files/00-config
+
+DEFAULT_REPO="https://github.com/KlipperScreen/KlipperScreen.git"
+DEFAULT_BRANCH=master
+
+KLIPPERSCREEN_REPO=${KLIPPERSCREEN_REPO:-$DEFAULT_REPO}
+KLIPPERSCREEN_BRANCH=${KLIPPERSCREEN_BRANCH:-$DEFAULT_BRANCH}
+
+export NETWORK=n
+export BACKEND=X
+export SERVICE=Y
+export START=0
+
+pushd /home/"${BASE_USER}" &> /dev/null || exit 1
+
+echo Cloning KlipperScreen Repo ${KLIPPERSCREEN_REPO} branch ${KLIPPERSCREEN_BRANCH} ..
+sudo -u "${BASE_USER}" git clone "${KLIPPERSCREEN_REPO}" KlipperScreen --branch "${KLIPPERSCREEN_BRANCH}"
+
+echo Running unattended KlipperScreen installation
+sudo -u "${BASE_USER}" "NETWORK=${NETWORK}" "BACKEND=${BACKEND}" "SERVICE=${SERVICE}" "START=${START}" ./KlipperScreen/scripts/KlipperScreen-install.sh
+
+
+popd &> /dev/null || exit 1
+
+# EOF 81-klipperscreen
\ No newline at end of file
diff --git a/configs/snapmakerj1/scripts/60-mainsailos.append b/configs/snapmakerj1/scripts/60-mainsailos.append
new file mode 100644
index 0000000..a4aed41
--- /dev/null
+++ b/configs/snapmakerj1/scripts/60-mainsailos.append
@@ -0,0 +1,5 @@
+# dirty cleanup
+#Not needed on snapmakerj1
+apt autoremove -y python3-opencv
+
+#EOF
\ No newline at end of file
diff --git a/configs/snapmakerj1/scripts/80-snapmakerj1 b/configs/snapmakerj1/scripts/80-snapmakerj1
new file mode 100644
index 0000000..ed4bed2
--- /dev/null
+++ b/configs/snapmakerj1/scripts/80-snapmakerj1
@@ -0,0 +1,67 @@
+#!/bin/bash
+set -xe
+export LC_ALL=C
+
+########################################
+# Functions and Basic Configuration    #
+########################################
+
+# Load common functions and set up cleanup trap
+source /common.sh
+install_cleanup_trap
+
+# Load additional configuration (e.g., BASE_USER)
+source /files/00-config
+
+# non existing serial 
+systemctl_if_exists mask serial-getty@ttyS0.service
+
+namei -l $PWD
+
+
+# fix display orientation
+[ ! -d /usr/share/X11/xorg.conf.d/ ] || mkdir -p /usr/share/X11/xorg.conf.d
+cat > /usr/share/X11/xorg.conf.d/90-display.conf << 'EOF'
+Section "Monitor"
+    Identifier "None-1"
+    # This identifier would be the same as the name of the connector printed by xrandr
+    # for example  "DVI-I-1 connected primary" means that the identifier is "DVI-I-1"
+    # another example "Unknown19-1 connected primary" some GPIO screens identify as Unknown19
+
+    Option "Rotate" "left"
+    # Valid rotation options are normal,inverted,left,right
+
+    #Option "PreferredMode" "1920x1080"
+    # May be necessary if you are not getting your preferred resolution.
+EndSection
+
+Section "InputClass"
+    Identifier "Coordinate Transformation Matrix"
+    MatchIsTouchscreen "on"
+    MatchDevicePath "/dev/input/event*"
+    MatchDriver "libinput"
+    Option "CalibrationMatrix" "0 -1 1  1 0 0  0 0 1"
+EndSection
+EOF
+
+# copy openocd configs
+
+
+# Setup printer_data
+ln -s /home/"${BASE_USER}"/klipper/config/printer-snapmaker-j1-hardware.cfg /home/"${BASE_USER}"/printer_data/config/hardware.cfg
+ln -s /home/"${BASE_USER}"/klipper/config/printer-snapmaker-j1-macros.cfg /home/"${BASE_USER}"/printer_data/config/macros.cfg
+cp /home/"${BASE_USER}"/klipper/config/printer-snapmaker-j1-printer.cfg /home/"${BASE_USER}"/printer_data/config/printer.cfg
+
+# Prepare OpenOCD 
+cp -r /files/openocd /home/"${BASE_USER}"/
+chown -h "${BASE_USER}:${BASE_USER}" /home/"${BASE_USER}"/printer_data/config/*.cfg
+
+
+# prepare the firmware
+pushd "/home/${BASE_USER}/klipper" &> /dev/null || exit 1
+sudo -u "${BASE_USER}" cp snapmakerj1.config .config
+sudo -u "${BASE_USER}" make olddefconfig
+sudo -u "${BASE_USER}" make
+popd &> /dev/null || exit 1
+
+# EOF 80-snapmakerj1
diff --git a/configs/snapmakerj1/scripts/files/00-config.append b/configs/snapmakerj1/scripts/files/00-config.append
new file mode 100644
index 0000000..25c7f86
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/00-config.append
@@ -0,0 +1,6 @@
+export KLIPPER_REPO="https://gitea.evilazrael.net/eazrael/klipper.git"
+export KLIPPER_BRANCH=snapmakerj1
+export DIST_NAME=SnapmakerJ1
+
+#nice, but this breaks the OS
+#export BASE_USER=j1
\ No newline at end of file
diff --git a/configs/snapmakerj1/scripts/files/openocd/gd32f307.cfg b/configs/snapmakerj1/scripts/files/openocd/gd32f307.cfg
new file mode 100644
index 0000000..f9604c9
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/gd32f307.cfg
@@ -0,0 +1,150 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# script for stm32f4x family
+
+#
+# stm32f4 devices support both JTAG and SWD transports.
+#
+source [find target/swj-dp.tcl]
+source [find mem_helper.tcl]
+
+if { [info exists CHIPNAME] } {
+   set _CHIPNAME $CHIPNAME
+} else {
+   set _CHIPNAME stm32f4x
+}
+
+set _ENDIAN little
+
+# Work-area is a space in RAM used for flash programming
+# By default use 32kB (Available RAM in smallest device STM32F410)
+if { [info exists WORKAREASIZE] } {
+   set _WORKAREASIZE $WORKAREASIZE
+} else {
+   set _WORKAREASIZE 0x8000
+}
+
+#jtag scan chain
+if { [info exists CPUTAPID] } {
+   set _CPUTAPID $CPUTAPID
+} else {
+   if { [using_jtag] } {
+      # See STM Document RM0090
+      # Section 38.6.3 - corresponds to Cortex-M4 r0p1
+      set _CPUTAPID 0x4ba00477
+   } {
+      set _CPUTAPID 0x2ba01477
+   }
+}
+
+swj_newdap $_CHIPNAME cpu -irlen 4 -ircapture 0x1 -irmask 0xf -expected-id $_CPUTAPID
+dap create $_CHIPNAME.dap -chain-position $_CHIPNAME.cpu
+
+if {[using_jtag]} {
+   jtag newtap $_CHIPNAME bs -irlen 5
+}
+
+set _TARGETNAME $_CHIPNAME.cpu
+target create $_TARGETNAME cortex_m -endian $_ENDIAN -dap $_CHIPNAME.dap
+
+$_TARGETNAME configure -work-area-phys 0x20000000 -work-area-size $_WORKAREASIZE -work-area-backup 0
+
+set _FLASHNAME $_CHIPNAME.flash
+flash bank $_FLASHNAME stm32f2x 0 0 0 0 $_TARGETNAME
+
+flash bank $_CHIPNAME.otp stm32f2x 0x1fff7800 0 0 0 $_TARGETNAME
+
+if { [info exists QUADSPI] && $QUADSPI } {
+   set a [llength [flash list]]
+   set _QSPINAME $_CHIPNAME.qspi
+   flash bank $_QSPINAME stmqspi 0x90000000 0 0 0 $_TARGETNAME 0xA0001000
+}
+
+# JTAG speed should be <= F_CPU/6. F_CPU after reset is 16MHz, so use F_JTAG = 2MHz
+#
+# Since we may be running of an RC oscilator, we crank down the speed a
+# bit more to be on the safe side. Perhaps superstition, but if are
+# running off a crystal, we can run closer to the limit. Note
+# that there can be a pretty wide band where things are more or less stable.
+#adapter speed 2000
+# the snapmaker j1 mainboard cannot take more
+adapter speed 50
+
+adapter srst delay 100
+if {[using_jtag]} {
+ jtag_ntrst_delay 100
+}
+
+reset_config srst_nogate
+
+if {![using_hla]} {
+   # if srst is not fitted use SYSRESETREQ to
+   # perform a soft reset
+   cortex_m reset_config sysresetreq
+}
+
+$_TARGETNAME configure -event examine-end {
+	# Enable debug during low power modes (uses more power)
+	# DBGMCU_CR |= DBG_STANDBY | DBG_STOP | DBG_SLEEP
+	mmw 0xE0042004 0x00000007 0
+
+	# Stop watchdog counters during halt
+	# DBGMCU_APB1_FZ |= DBG_IWDG_STOP | DBG_WWDG_STOP
+	mmw 0xE0042008 0x00001800 0
+}
+
+tpiu create $_CHIPNAME.tpiu -dap $_CHIPNAME.dap -ap-num 0 -baseaddr 0xE0040000
+
+lappend _telnet_autocomplete_skip _proc_pre_enable_$_CHIPNAME.tpiu
+proc _proc_pre_enable_$_CHIPNAME.tpiu {_chipname} {
+	targets $_chipname.cpu
+
+	if { [$_chipname.tpiu cget -protocol] eq "sync" } {
+		switch [$_chipname.tpiu cget -port-width] {
+			1 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 1 bit; GPIOE[2-3] to AF0
+				mmw 0xE0042004 0x00000060 0x000000c0
+				mmw 0x40021020 0x00000000 0x0000ff00
+				mmw 0x40021000 0x000000a0 0x000000f0
+				mmw 0x40021008 0x000000f0 0x00000000
+			  }
+			2 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 2 bit; GPIOE[2-4] to AF0
+				mmw 0xE0042004 0x000000a0 0x000000c0
+				mmw 0x40021020 0x00000000 0x000fff00
+				mmw 0x40021000 0x000002a0 0x000003f0
+				mmw 0x40021008 0x000003f0 0x00000000
+			  }
+			4 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 4 bit; GPIOE[2-6] to AF0
+				mmw 0xE0042004 0x000000e0 0x000000c0
+				mmw 0x40021020 0x00000000 0x0fffff00
+				mmw 0x40021000 0x00002aa0 0x00003ff0
+				mmw 0x40021008 0x00003ff0 0x00000000
+			  }
+		}
+	} else {
+		# Set TRACE_IOEN; TRACE_MODE to async
+		mmw 0xE0042004 0x00000020 0x000000c0
+	}
+}
+
+$_CHIPNAME.tpiu configure -event pre-enable "_proc_pre_enable_$_CHIPNAME.tpiu $_CHIPNAME"
+
+$_TARGETNAME configure -event reset-init {
+	# Configure PLL to boost clock to HSI x 4 (64 MHz)
+	mww 0x40023804 0x08012008   ;# RCC_PLLCFGR 16 Mhz /8 (M) * 128 (N) /4(P)
+	mww 0x40023C00 0x00000102   ;# FLASH_ACR = PRFTBE | 2(Latency)
+	mmw 0x40023800 0x01000000 0 ;# RCC_CR |= PLLON
+	sleep 10                    ;# Wait for PLL to lock
+	mmw 0x40023808 0x00001000 0 ;# RCC_CFGR |= RCC_CFGR_PPRE1_DIV2
+	mmw 0x40023808 0x00000002 0 ;# RCC_CFGR |= RCC_CFGR_SW_PLL
+
+	# Boost JTAG frequency
+	adapter speed 8000
+}
+
+$_TARGETNAME configure -event reset-start {
+	# Reduce speed since CPU speed will slow down to 16MHz with the reset
+	adapter speed 2000
+}
diff --git a/configs/snapmakerj1/scripts/files/openocd/stlink-dap.cfg b/configs/snapmakerj1/scripts/files/openocd/stlink-dap.cfg
new file mode 100644
index 0000000..5c24cbd
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/stlink-dap.cfg
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+#
+# STMicroelectronics ST-LINK/V1, ST-LINK/V2, ST-LINK/V2-1, STLINK-V3 in-circuit
+# debugger/programmer
+#
+# This new interface driver creates a ST-Link wrapper for ARM-DAP named "dapdirect"
+# Old ST-LINK/V1 and ST-LINK/V2 pre version V2J24 don't support "dapdirect"
+#
+# SWIM transport is natively supported
+#
+
+adapter driver st-link
+st-link vid_pid 0x0483 0x3744 0x0483 0x3748 0x0483 0x374b 0x0483 0x374d 0x0483 0x374e 0x0483 0x374f 0x0483 0x3752 0x0483 0x3753 0x0483 0x3754
+
+# transport select dapdirect_jtag
+# transport select dapdirect_swd
+# transport select swim
+
+# Optionally specify the serial number of usb device
+# e.g.
+# adapter serial "\xaa\xbc\x6e\x06\x50\x75\xff\x55\x17\x42\x19\x3f"
diff --git a/configs/snapmakerj1/scripts/files/openocd/stlink.cfg b/configs/snapmakerj1/scripts/files/openocd/stlink.cfg
new file mode 100644
index 0000000..e4906b7
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/stlink.cfg
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+#
+# STMicroelectronics ST-LINK/V1, ST-LINK/V2, ST-LINK/V2-1, STLINK-V3 in-circuit
+# debugger/programmer
+#
+
+adapter driver hla
+hla_layout stlink
+hla_device_desc "ST-LINK"
+hla_vid_pid 0x0483 0x3744 0x0483 0x3748 0x0483 0x374b 0x0483 0x374d 0x0483 0x374e 0x0483 0x374f 0x0483 0x3752 0x0483 0x3753 0x0483 0x3754
+
+# Optionally specify the serial number of ST-LINK/V2 usb device.  ST-LINK/V2
+# devices seem to have serial numbers with unreadable characters.  ST-LINK/V2
+# firmware version >= V2.J21.S4 recommended to avoid issues with adapter serial
+# number reset issues.
+# eg.
+# adapter serial "\xaa\xbc\x6e\x06\x50\x75\xff\x55\x17\x42\x19\x3f"
diff --git a/configs/snapmakerj1/scripts/files/openocd/stm32f1x.cfg b/configs/snapmakerj1/scripts/files/openocd/stm32f1x.cfg
new file mode 100644
index 0000000..a22cc09
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/stm32f1x.cfg
@@ -0,0 +1,99 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# script for stm32f1x family
+
+#
+# stm32 devices support both JTAG and SWD transports.
+#
+source [find target/swj-dp.tcl]
+source [find mem_helper.tcl]
+
+if { [info exists CHIPNAME] } {
+   set _CHIPNAME $CHIPNAME
+} else {
+   set _CHIPNAME stm32f1x
+}
+
+set _ENDIAN little
+
+# Work-area is a space in RAM used for flash programming
+# By default use 4kB (as found on some STM32F100s)
+if { [info exists WORKAREASIZE] } {
+   set _WORKAREASIZE $WORKAREASIZE
+} else {
+   set _WORKAREASIZE 0x1000
+}
+
+# Allow overriding the Flash bank size
+if { [info exists FLASH_SIZE] } {
+    set _FLASH_SIZE $FLASH_SIZE
+} else {
+    # autodetect size
+    set _FLASH_SIZE 0x100000
+}
+
+#jtag scan chain
+if { [info exists CPUTAPID] } {
+   set _CPUTAPID $CPUTAPID
+set _CPUTAPID } else {
+   if { [using_jtag] } {
+      # See STM Document RM0008 Section 26.6.3
+      set _CPUTAPID 0x3ba00477
+   } {
+      # this is the SW-DP tap id not the jtag tap id
+      #set _CPUTAPID 0x1ba01477
+      set _CPUTAPID 0x2ba01477
+   }
+}
+
+swj_newdap $_CHIPNAME cpu -irlen 4 -ircapture 0x1 -irmask 0xf -expected-id $_CPUTAPID
+dap create $_CHIPNAME.dap -chain-position $_CHIPNAME.cpu
+
+if {[using_jtag]} {
+   jtag newtap $_CHIPNAME bs -irlen 5
+}
+
+set _TARGETNAME $_CHIPNAME.cpu
+target create $_TARGETNAME cortex_m -endian $_ENDIAN -dap $_CHIPNAME.dap
+
+$_TARGETNAME configure -work-area-phys 0x20000000 -work-area-size $_WORKAREASIZE -work-area-backup 0
+
+# flash size will be probed
+set _FLASHNAME $_CHIPNAME.flash
+flash bank $_FLASHNAME stm32f1x 0x08000000 $_FLASH_SIZE 0 0 $_TARGETNAME
+
+# JTAG speed should be <= F_CPU/6. F_CPU after reset is 8MHz, so use F_JTAG = 1MHz
+adapter speed 50
+
+adapter srst delay 100
+if {[using_jtag]} {
+ jtag_ntrst_delay 100
+}
+
+reset_config srst_nogate
+
+if {![using_hla]} {
+    # if srst is not fitted use SYSRESETREQ to
+    # perform a soft reset
+    cortex_m reset_config sysresetreq
+}
+
+$_TARGETNAME configure -event examine-end {
+	# DBGMCU_CR |= DBG_WWDG_STOP | DBG_IWDG_STOP |
+	#              DBG_STANDBY | DBG_STOP | DBG_SLEEP
+	mmw 0xE0042004 0x00000307 0
+}
+
+tpiu create $_CHIPNAME.tpiu -dap $_CHIPNAME.dap -ap-num 0 -baseaddr 0xE0040000
+
+lappend _telnet_autocomplete_skip _proc_pre_enable_$_CHIPNAME.tpiu
+proc _proc_pre_enable_$_CHIPNAME.tpiu {_targetname} {
+	targets $_targetname
+
+	# Set TRACE_IOEN; TRACE_MODE is set to async; when using sync
+	# change this value accordingly to configure trace pins
+	# assignment
+	mmw 0xE0042004 0x00000020 0
+}
+
+$_CHIPNAME.tpiu configure -event pre-enable "_proc_pre_enable_$_CHIPNAME.tpiu $_TARGETNAME"
diff --git a/configs/snapmakerj1/scripts/files/openocd/stm32f3x.cfg b/configs/snapmakerj1/scripts/files/openocd/stm32f3x.cfg
new file mode 100644
index 0000000..4ecc7ed
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/stm32f3x.cfg
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# script for stm32f3x family
+
+#
+# stm32 devices support both JTAG and SWD transports.
+#
+source [find target/swj-dp.tcl]
+source [find mem_helper.tcl]
+
+if { [info exists CHIPNAME] } {
+   set _CHIPNAME $CHIPNAME
+} else {
+   set _CHIPNAME stm32f3x
+}
+
+set _ENDIAN little
+
+# Work-area is a space in RAM used for flash programming
+# By default use 16kB
+if { [info exists WORKAREASIZE] } {
+   set _WORKAREASIZE $WORKAREASIZE
+} else {
+   set _WORKAREASIZE 0x4000
+}
+
+# JTAG speed should be <= F_CPU/6. F_CPU after reset is 8MHz, so use F_JTAG = 1MHz
+#
+# Since we may be running of an RC oscilator, we crank down the speed a
+# bit more to be on the safe side. Perhaps superstition, but if are
+# running off a crystal, we can run closer to the limit. Note
+# that there can be a pretty wide band where things are more or less stable.
+adapter speed 1000
+
+adapter srst delay 100
+if {[using_jtag]} {
+ jtag_ntrst_delay 100
+}
+
+#jtag scan chain
+if { [info exists CPUTAPID] } {
+   set _CPUTAPID $CPUTAPID
+} else {
+   if { [using_jtag] } {
+      # See STM Document RM0316
+      # Section 29.6.3 - corresponds to Cortex-M4 r0p1
+      set _CPUTAPID 0x4ba00477
+   } {
+      set _CPUTAPID 0x2ba01477
+   }
+}
+
+swj_newdap $_CHIPNAME cpu -irlen 4 -ircapture 0x1 -irmask 0xf -expected-id $_CPUTAPID
+dap create $_CHIPNAME.dap -chain-position $_CHIPNAME.cpu
+
+if {[using_jtag]} {
+   jtag newtap $_CHIPNAME bs -irlen 5
+}
+
+set _TARGETNAME $_CHIPNAME.cpu
+target create $_TARGETNAME cortex_m -endian $_ENDIAN -dap $_CHIPNAME.dap
+
+$_TARGETNAME configure -work-area-phys 0x20000000 -work-area-size $_WORKAREASIZE -work-area-backup 0
+
+set _FLASHNAME $_CHIPNAME.flash
+flash bank $_FLASHNAME stm32f1x 0 0 0 0 $_TARGETNAME
+
+reset_config srst_nogate
+
+if {![using_hla]} {
+   # if srst is not fitted use SYSRESETREQ to
+   # perform a soft reset
+   cortex_m reset_config sysresetreq
+}
+
+proc stm32f3x_default_reset_start {} {
+	# Reset clock is HSI (8 MHz)
+	adapter speed 1000
+}
+
+proc stm32f3x_default_examine_end {} {
+	# Enable debug during low power modes (uses more power)
+	mmw 0xe0042004 0x00000007 0 ;# DBGMCU_CR |= DBG_STANDBY | DBG_STOP | DBG_SLEEP
+
+	# Stop watchdog counters during halt
+	mmw 0xe0042008 0x00001800 0 ;# DBGMCU_APB1_FZ |= DBG_IWDG_STOP | DBG_WWDG_STOP
+}
+
+proc stm32f3x_default_reset_init {} {
+	# Configure PLL to boost clock to HSI x 8 (64 MHz)
+	mww 0x40021004 0x00380400   ;# RCC_CFGR = PLLMUL[3:1] | PPRE1[2]
+	mmw 0x40021000 0x01000000 0 ;# RCC_CR |= PLLON
+	mww 0x40022000 0x00000012   ;# FLASH_ACR = PRFTBE | LATENCY[1]
+	sleep 10                    ;# Wait for PLL to lock
+	mmw 0x40021004 0x00000002 0 ;# RCC_CFGR |= SW[1]
+
+	# Boost JTAG frequency
+	adapter speed 8000
+}
+
+# Default hooks
+$_TARGETNAME configure -event examine-end { stm32f3x_default_examine_end }
+$_TARGETNAME configure -event reset-start { stm32f3x_default_reset_start }
+$_TARGETNAME configure -event reset-init { stm32f3x_default_reset_init }
+
+tpiu create $_CHIPNAME.tpiu -dap $_CHIPNAME.dap -ap-num 0 -baseaddr 0xE0040000
+
+lappend _telnet_autocomplete_skip _proc_pre_enable_$_CHIPNAME.tpiu
+proc _proc_pre_enable_$_CHIPNAME.tpiu {_targetname} {
+	targets $_targetname
+
+	# Set TRACE_IOEN; TRACE_MODE is set to async; when using sync
+	# change this value accordingly to configure trace pins
+	# assignment
+	mmw 0xe0042004 0x00000020 0
+}
+
+$_CHIPNAME.tpiu configure -event pre-enable "_proc_pre_enable_$_CHIPNAME.tpiu $_TARGETNAME"
diff --git a/configs/snapmakerj1/scripts/files/openocd/stm32f4x.cfg b/configs/snapmakerj1/scripts/files/openocd/stm32f4x.cfg
new file mode 100644
index 0000000..0b4bf34
--- /dev/null
+++ b/configs/snapmakerj1/scripts/files/openocd/stm32f4x.cfg
@@ -0,0 +1,148 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+# script for stm32f4x family
+
+#
+# stm32f4 devices support both JTAG and SWD transports.
+#
+source [find target/swj-dp.tcl]
+source [find mem_helper.tcl]
+
+if { [info exists CHIPNAME] } {
+   set _CHIPNAME $CHIPNAME
+} else {
+   set _CHIPNAME stm32f4x
+}
+
+set _ENDIAN little
+
+# Work-area is a space in RAM used for flash programming
+# By default use 32kB (Available RAM in smallest device STM32F410)
+if { [info exists WORKAREASIZE] } {
+   set _WORKAREASIZE $WORKAREASIZE
+} else {
+   set _WORKAREASIZE 0x8000
+}
+
+#jtag scan chain
+if { [info exists CPUTAPID] } {
+   set _CPUTAPID $CPUTAPID
+} else {
+   if { [using_jtag] } {
+      # See STM Document RM0090
+      # Section 38.6.3 - corresponds to Cortex-M4 r0p1
+      set _CPUTAPID 0x4ba00477
+   } {
+      set _CPUTAPID 0x2ba01477
+   }
+}
+
+swj_newdap $_CHIPNAME cpu -irlen 4 -ircapture 0x1 -irmask 0xf -expected-id $_CPUTAPID
+dap create $_CHIPNAME.dap -chain-position $_CHIPNAME.cpu
+
+if {[using_jtag]} {
+   jtag newtap $_CHIPNAME bs -irlen 5
+}
+
+set _TARGETNAME $_CHIPNAME.cpu
+target create $_TARGETNAME cortex_m -endian $_ENDIAN -dap $_CHIPNAME.dap
+
+$_TARGETNAME configure -work-area-phys 0x20000000 -work-area-size $_WORKAREASIZE -work-area-backup 0
+
+set _FLASHNAME $_CHIPNAME.flash
+flash bank $_FLASHNAME stm32f2x 0 0 0 0 $_TARGETNAME
+
+flash bank $_CHIPNAME.otp stm32f2x 0x1fff7800 0 0 0 $_TARGETNAME
+
+if { [info exists QUADSPI] && $QUADSPI } {
+   set a [llength [flash list]]
+   set _QSPINAME $_CHIPNAME.qspi
+   flash bank $_QSPINAME stmqspi 0x90000000 0 0 0 $_TARGETNAME 0xA0001000
+}
+
+# JTAG speed should be <= F_CPU/6. F_CPU after reset is 16MHz, so use F_JTAG = 2MHz
+#
+# Since we may be running of an RC oscilator, we crank down the speed a
+# bit more to be on the safe side. Perhaps superstition, but if are
+# running off a crystal, we can run closer to the limit. Note
+# that there can be a pretty wide band where things are more or less stable.
+adapter speed 50
+
+adapter srst delay 100
+if {[using_jtag]} {
+ jtag_ntrst_delay 100
+}
+
+reset_config srst_nogate
+
+if {![using_hla]} {
+   # if srst is not fitted use SYSRESETREQ to
+   # perform a soft reset
+   cortex_m reset_config sysresetreq
+}
+
+$_TARGETNAME configure -event examine-end {
+	# Enable debug during low power modes (uses more power)
+	# DBGMCU_CR |= DBG_STANDBY | DBG_STOP | DBG_SLEEP
+	mmw 0xE0042004 0x00000007 0
+
+	# Stop watchdog counters during halt
+	# DBGMCU_APB1_FZ |= DBG_IWDG_STOP | DBG_WWDG_STOP
+	mmw 0xE0042008 0x00001800 0
+}
+
+tpiu create $_CHIPNAME.tpiu -dap $_CHIPNAME.dap -ap-num 0 -baseaddr 0xE0040000
+
+lappend _telnet_autocomplete_skip _proc_pre_enable_$_CHIPNAME.tpiu
+proc _proc_pre_enable_$_CHIPNAME.tpiu {_chipname} {
+	targets $_chipname.cpu
+
+	if { [$_chipname.tpiu cget -protocol] eq "sync" } {
+		switch [$_chipname.tpiu cget -port-width] {
+			1 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 1 bit; GPIOE[2-3] to AF0
+				mmw 0xE0042004 0x00000060 0x000000c0
+				mmw 0x40021020 0x00000000 0x0000ff00
+				mmw 0x40021000 0x000000a0 0x000000f0
+				mmw 0x40021008 0x000000f0 0x00000000
+			  }
+			2 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 2 bit; GPIOE[2-4] to AF0
+				mmw 0xE0042004 0x000000a0 0x000000c0
+				mmw 0x40021020 0x00000000 0x000fff00
+				mmw 0x40021000 0x000002a0 0x000003f0
+				mmw 0x40021008 0x000003f0 0x00000000
+			  }
+			4 {
+				# Set TRACE_IOEN; TRACE_MODE to sync 4 bit; GPIOE[2-6] to AF0
+				mmw 0xE0042004 0x000000e0 0x000000c0
+				mmw 0x40021020 0x00000000 0x0fffff00
+				mmw 0x40021000 0x00002aa0 0x00003ff0
+				mmw 0x40021008 0x00003ff0 0x00000000
+			  }
+		}
+	} else {
+		# Set TRACE_IOEN; TRACE_MODE to async
+		mmw 0xE0042004 0x00000020 0x000000c0
+	}
+}
+
+$_CHIPNAME.tpiu configure -event pre-enable "_proc_pre_enable_$_CHIPNAME.tpiu $_CHIPNAME"
+
+$_TARGETNAME configure -event reset-init {
+	# Configure PLL to boost clock to HSI x 4 (64 MHz)
+	mww 0x40023804 0x08012008   ;# RCC_PLLCFGR 16 Mhz /8 (M) * 128 (N) /4(P)
+	mww 0x40023C00 0x00000102   ;# FLASH_ACR = PRFTBE | 2(Latency)
+	mmw 0x40023800 0x01000000 0 ;# RCC_CR |= PLLON
+	sleep 10                    ;# Wait for PLL to lock
+	mmw 0x40023808 0x00001000 0 ;# RCC_CFGR |= RCC_CFGR_PPRE1_DIV2
+	mmw 0x40023808 0x00000002 0 ;# RCC_CFGR |= RCC_CFGR_SW_PLL
+
+	# Boost JTAG frequency
+	adapter speed 50
+}
+
+$_TARGETNAME configure -event reset-start {
+	# Reduce speed since CPU speed will slow down to 16MHz with the reset
+	adapter speed 50
+}
diff --git a/modules/armbian/10-armbian b/modules/armbian/10-armbian
index adb26f4..07b456d 100644
--- a/modules/armbian/10-armbian
+++ b/modules/armbian/10-armbian
@@ -25,7 +25,7 @@ if_group_exists_run() {
 ########################################
 
 BASE_PASSWORD=armbian
-DEPS=(armbian-config avahi-daemon bash-completion iptables)
+DEPS=(armbian-config avahi-daemon bash-completion iptables strace)
 
 apt-get install --yes "${DEPS[@]}"
 
diff --git a/modules/generic/00-upgrade b/modules/generic/00-upgrade
index 4096048..af95476 100644
--- a/modules/generic/00-upgrade
+++ b/modules/generic/00-upgrade
@@ -32,7 +32,15 @@ fi
 # Execute upgrade                      #
 ########################################
 
+# Armbian firmware is broken right now .. 
+# Unpacking armbian-firmware (26.2.1) over (25.11.0-trunk) ...
+# dpkg: error processing archive /tmp/apt-dpkg-install-xpkFLI/08-armbian-firmware_26.2.1_all.deb (--unpack):
+# unable to open '/lib/firmware/qcom/sm8550/ayn/odin2mini/a740_zap.mbn.dpkg-new': No such file or directory
+# and it gets removed later anyway...
+#apt-mark hold armbian-firmware
+apt-get -y remove armbian-firmware
 apt-get -y -o Dpkg::Options::="--force-confnew" upgrade
+#apt-mark unhold armbian-firmware
 
 ########################################
 # Cleanup                              #
diff --git a/modules/generic/50-klipper b/modules/generic/50-klipper
index 6de4b5e..e07b21e 100644
--- a/modules/generic/50-klipper
+++ b/modules/generic/50-klipper
@@ -13,6 +13,8 @@ install_cleanup_trap
 # Load additional configuration (e.g., BASE_USER)
 source /files/00-config
 
+set
+
 ########################################
 # Configuration                        #
 ########################################
@@ -22,6 +24,9 @@ ENV_DIR="/home/${BASE_USER}/klippy-env"
 
 REPO="https://github.com/Klipper3d/klipper.git"
 
+KLIPPER_REPO=${KLIPPER_REPO:-$REPO}
+KLIPPER_BRANCH=${KLIPPER_BRANCH:-"master"}
+
 DEPS=(git \
       # Packages for python cffi
       virtualenv python3-dev libffi-dev build-essential \
@@ -64,8 +69,8 @@ pushd "/home/${BASE_USER}" &> /dev/null || exit 1
 # Clone repository                     #
 ########################################
 
-echo "Cloning repo ..."
-sudo -u "${BASE_USER}" git clone ${REPO} klipper
+echo "Cloning repo ${KLIPPER_REPO} branch ${KLIPPER_BRANCH}..."
+sudo -u "${BASE_USER}" git clone ${KLIPPER_REPO} klipper --branch ${KLIPPER_BRANCH}
 
 ########################################
 # Create printer_data directories      #
diff --git a/modules/generic/99-unhold-packages b/modules/generic/99-unhold-packages
index e8e16dd..3079208 100644
--- a/modules/generic/99-unhold-packages
+++ b/modules/generic/99-unhold-packages
@@ -18,3 +18,5 @@ source /files/00-config
 ########################################
 
 #apt-mark unhold "${PKGS[@]}" > /dev/null 2>&1 || true
+
+fstrim -v / || true
